lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -289,6 +294,110 @@ def filter_query(filters: list[str], query: Query) -> Query:
 
         return query.filter(and_(*and_list))
 
+    def accumulate(self, aggregate: JSONOBJECT, key: str, value: Any):
+        """Recursive helper to accumulate the metadata namespace
+
+        Iterate through a list of metadata key/value pairs to construct a
+        hierarchical aggregation of all metadata keys across the selected
+        datasets. Each key in the hierarchy is represented as a key in a
+        nested JSON object. "Leaf" keys have the value None. E.g.,
+
+            {
+                "dataset": {"name": None, "metalog": {"pbench": {"script": None}}},
+                "server": {"deletion": None, "tarball-path": None},
+                "global": {"server": {"legacy": {"sha1": None}}}
+            }
+
+        Args:
+            aggregate: a JSONOBJECT to update with the recursive key/value
+            key: the current metadata key path element
+            value: the current metadata key's value
+        """
+        if isinstance(value, dict):
+            p = aggregate.get(key)
+            if p is None:
+                p = {}
+                aggregate[key] = p
+            for k, v in value.items():
+                self.accumulate(p, k, v)
+        elif key not in aggregate:
+            aggregate[key] = None
+
+    def keyspace(self, query: Query) -> JSONOBJECT:
+        """Aggregate the dataset metadata keyspace
+
+        Run the query we've compiled, but instead of returning Dataset proxies,
+        we only want the metadata key/value pairs we've selected.
+
+        NOTE: The SQL left outer join returns a row for each row in the "left"
+        table (Dataset) even if there is no matching foreign key in the "right"
+        table (Metadata). This means a dataset with no metadata will result in
+        a join row here with key and value of None. The `elif` in the loop will
+        silently ignore rows with a null key to handle this case.
+
+        Args:
+            query: The basic filtered SQLAlchemy query object
+
+        Returns:
+            The aggregated keyspace JSON object
+        """
+        aggregate: JSONOBJECT = {
+            "dataset": {c.name: None for c in Dataset.__table__._columns}
+        }
+        list = query.with_entities(Metadata.key, Metadata.value).all()
+        for k, v in list:
+            # "metalog" is a top-level key in the Metadata schema, but we
+            # report it as a sub-key of "dataset".
```

Comment:
Okay, a bit confused, why did you choose to add it as a sub-key of the dataset? 

Also, I realized I think we should send the `metalog` only if asked explicitly, the output can get big quickly for complex and long benchmark runs. Thought?

Reply:
> Okay, a bit confused, why did you choose to add it as a sub-key of the dataset?

Actually I originally wanted to make it a column of `Dataset` because it really is part of the "identity" of the dataset rather than external tracking context like the `server` namespace. At the time I added it, I felt that I didn't want to change the DB schema. And at this point I'd feel obligated to write, and somehow test, migration code if I moved it, so I probably won't. 😆 

> Also, I realized I think we should send the `metalog` only if asked explicitly, the output can get big quickly for complex and long benchmark runs. Thought?

The `metadata.log` is an important part of the test run, including the script name, config data, and such. The problems here are (1) the weird key names that the agent postprocessing generates, and (2) the extremely "irregular" key space depending on the benchmark and parameters; but I think it's important and valuable to expose all this for display and filtering.

What I *did* wrestle with was whether to arbitrarily suppress the `server.index-map`. While it's not so bad here since we only represent the keys, the *value* (if someone decided to enable display) can be enormous.